charms/kfp-metadata-writer/src/components/k8s_service_info_component.py

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+# Copyright 2024 Ubuntu
+# See LICENSE file for licensing details.
+
+from typing import List, Optional, Union
+
+from charmed_kubeflow_chisme.components.component import Component
+from charms.mlops_libs.v0.k8s_service_info import (
+    KubernetesServiceInfoObject,
+    KubernetesServiceInfoRelationDataMissingError,
+    KubernetesServiceInfoRelationMissingError,
+    KubernetesServiceInfoRequirer,
+)
+from ops import ActiveStatus, BlockedStatus, CharmBase, StatusBase
+from ops.framework import BoundEvent
+
+
+class K8sServiceInfoComponent(Component):
+    """A Component that wraps the requirer side of the k8s_service_info charm library.
+
+    Args:
+        charm(CharmBase): the requirer charm
+        relation_name(str, Optional): name of the relation that uses the k8s-service interface
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: Optional[str] = "k8s-service-info",
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+    ):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.charm = charm
+        self.refresh_event = refresh_event
+
+        self._k8s_service_info_requirer = KubernetesServiceInfoRequirer(
+            charm=self.charm, relation_name=self.relation_name, refresh_event=self.refresh_event
+        )
+
+    def get_service_info(self) -> KubernetesServiceInfoObject:
+        """Wrap the get_data method and return a KubernetesServiceInfoObject."""
+        return self._k8s_service_info_requirer.get_data()
+
+    def get_status(self) -> StatusBase:
+        """Return this component's status based on the presence of the relation and its data."""
+        try:
+            self.get_service_info()
+        except KubernetesServiceInfoRelationMissingError as rel_error:
+            return BlockedStatus(f"{rel_error.message} Please add the missing relation.")
+        except KubernetesServiceInfoRelationDataMissingError as data_error:
+            # Nothing can be done, just re-raise
+            raise data_error
```

Comment:
I worry this is too harsh.  This will happen transiently when a relation is established [^1].  Can we add some explanation to this saying it can be transient, but if it persists then it is an error?

[^1]: because [`relation-changed` always follows a `relation-joined`, so there will be some time between relation-joined and when the other side actually posts their data (triggering a second `relation-changed`)

Reply:
Should we return a `WaitingStatus`?

While I agree this could be too harsh, I worry that we report back a status when we should report back an error. I have seen in other charms we do this, and most of the times it is not very obvious what is going on. I'm open to suggestions.

Reply:
I suggest `WaitingStatus` with a message making it clear that, if `WaitingStatus` persists, its probably an error.  So like `WaitingStatus("No data found in X relation.  This may be transient, but if it persists it is likely an error") or similar

Reply:
Personally I do not like that message, I have seen many charms that use the charmReconciler hidding actual issues behind this message and for someone who has to debug charms, it is a pain to spot the actual issue. I can change it to what you are suggesting, but knowing this is a behaviour we should probably change.

Reply:
Updated, let me know what you think